### ISSUE TITLE
More than a proof-of-concept?

### ISSUE THREAD

**Issue by morrisonlevi** [spatie/typed#26]
> This package is a mere proof of concept about what's possible in PHP's userland to improve type checking.

Any plans of making this more than a proof-of-concept?

**Comment by brendt** [comment]
We've been talking about how this package could be usable, yes. There are two main concerns:

- IDE autocompletion. Now I know that many PHP devs still don't rely on IDE support, but to me it's a must.
- Performance. You probably already looked at the source code, and know we're using `__invoke` on type classes to validate types. I haven't done any benchmarks yet, but there is an obvious overhead.

Now on to solutions for these problems:

Autocompletion can be solved by adding a generator to this package. It will generate concrete implementations of data structure that you can extend. So instead of `new List(T::generic(Post::class))`; the package could generate `PostList`. These generated classes a properly documented so the IDE knows what it's dealing with. I've already got a working proof of concept for all data structures provided by the package, though only tuples are not a 100% accurate. It's a matter of taking the time and writing a generator. There are some design decisions there, but nothing difficult.

On the topic of performance, I should first add some benchmarks to be able to measure changes. I believe switching to a normal function call instead of `__invoke` might already improve the situation, but I should definitely check for how this performs on large data sets etc.

So what's your opinion? Are you asking the question for a specific reason? If there's a real interest from people, I will consider investing time in this. I for one would like to use the package myself in some side-projects, and outsiders interest would be good motivation.

**Comment by morrisonlevi** [comment]
Which IDEs do you are about?

There are various projects researching this space and perhaps we could benefit from more collaboration. The goals and designs differ a bit from project to project but they are all exploring userland type checking to full the void of generics.

**Comment by brendt** [comment]
Say you have the following:

```
$line = new Tuple(T::generic(Point::class), T::generic(Point::class));
```

I want PHPStorm to be able to autocomplete:

```
$line[0]->... // Point's methods etc.
$line[1]->... 
```

> There are various projects researching this space and perhaps we could benefit from more collaboration

That's an interesting thought. Do you think that there are actual production use-cases for userland type checking like these? In my experience stricter type checking get most of its benefit during development in the form of errors showing up much more soon, and autocompletion. 

So while it's an interesting thought experiment, I don't know when we'd use it in this form in production projects. I think we briefly discussed this on Reddit a while back, I personally believe there's much more value in adding syntax that can be statically analysed, without having runtime checks. 

PHP is an odd language in this regard, there are some type checks at runtime, but it can never reach the safety that statically, strong typed languages guarantee. I tend to compare this a lot to Rust's type system.

Nevertheless: runtime type checks in PHP have their merit, but I cannot think of any real use cases for the projects I work on, if there's no autocompletion and IDE/static error detection.

**Comment by morrisonlevi** [comment]
Obviously getting generics into PHP itself is the best solution; no argument there.

To support this in PhpStorm I would suggest a library plugin. I don't know anything about these other than it's possible to teach PhpStorm library-specific things it can't infer automatically. This is probably only worth it if the project is stable and you are committed to it.

My library does type checking through `assert` statements - this allows people to turn the type checking on and off. That does help production a bit.

**Comment by Niko9911** [comment]
@brendt When you were talking about performance problems, could zephir-lang.com or php-cpp.com be helpful in these manners?

**Comment by brendt** [comment]
@Niko9911 wouldn't that mean writing a strict type system as a real extension? I'm not acquainted enough with C or extension development to really get something going, there are people much more able to tackle that task 😄

Also: this is a userland implementation, doing some piggy-back riding on the existing type system in PHP; writing a real type system is something quiet different.

**Comment by Niko9911** [comment]
@brendt Yes it would and that's also point of that.

This is just simple work around until they will implement this into the core PHP, tough I think they will not. Due PHP type system is too large to handle and I don't have required knowledge about the PHP core source code on this manner, I could start writing small C extension as a work around to implement this feature as C extension. Would be probably performing little bit better at least, but will not be good way to do it tough...

Dunno. Best option would be finding PHP Core dev to do this =P

**Comment by brendt** [comment]
I share your opinion that the chances of getting a good strict type system in PHP are very small. Writing this package (or another one) in Zephir would definitely be possible, but than people have to modify their PHP config and add extensions. 

The more I think about it, the less merit I see for a better type system in PHP's current eco system. A better type system in itself doesn't really serve anything. It's the tooling around the system that gives it its power, not the system itself.

In other words: there's no need for strong types if you can only check them at runtime, generics loose most of their advantages when you don't have the proper static analysis to interpret them.

**Comment by Niko9911** [comment]
Zephir is good, but the problem is the parser. After Zephir is compiled to C, the code quality is bad. Okey, you get the job done, but it is not the correct way to go. If you go with PHP-CPP you will retain "Access" to the core. With Zephir you're just relying the transpiler to create the C code used in the extension. I have my own doubts about it. Please see https://github.com/phalcon/zephir/issues/694.

I agree, it's just a tool. But using this tool you will improve overall code quality and you will find less weird bugs from your code. Yup, you can build checks to check if type is correct, but it would be much easier if yo have those check implemented in the PHP as C library, when it would be performing much better.

I will be starting to write this into C extension as a "Hobby Project". I can't promise any output, but with this way we can get some attention from the PHP core devs and if this get's enough attention it will be implemented into PHP in version 9.xxx with a good luck.

PHP extension might sound dumm idea, like you have to compile it and etc... But if you take look into PHP extension world, it isn't actually so abstract idea to create PHP extension. Just take look into PHP ds. ;)

**Comment by brendt** [comment]
I'm actually using this in one of my projects now, so we'll see where it goes.
